sdk/core/core/inc/az_span.h

```diff
@@ -84,30 +88,37 @@ AZ_NODISCARD AZ_INLINE int32_t az_span_capacity(az_span span) { return span._int
   }
 
 /**
- * @brief Creates a az_span expression. For example,
- * `some_function(AZ_SPAN_FROM_STR("Hello world"));`
- * where
- * `void some_function(const az_span span);`
+ * @brief The AZ_SPAN_FROM_STR macro returns an az_span expression over a literal string. For
+ * example: `some_function(AZ_SPAN_FROM_STR("Hello world"));` where `void some_function(const
+ * az_span span);`
  */
 #define AZ_SPAN_FROM_STR(STRING_LITERAL) (az_span) AZ_SPAN_LITERAL_FROM_STR(STRING_LITERAL)
 
 /**
- * @brief Init span
+ * @brief az_span_init returns a span over a byte buffer.
  *
+ * @param[in] ptr The memory address of the 1st byte in the byte buffer
+ * @param[in] length The number of bytes initialized in the byte buffer
+ * @param[in] capacity The number of total bytes in the byte buffer
+ * @return az_span The az_span "view" over the byte buffer
  */
 AZ_NODISCARD az_span az_span_init(uint8_t* ptr, int32_t length, int32_t capacity);
 
 /**
- * @brief Creates a span from zero terminated string
+ * @brief az_span_from_str returns an az_span from a 0-terminated array of bytes (chars)
  *
+ * @param[in] str The pointer to the 0-terminated array of bytes (chars)
+ * @return az_span An az_span over the byte buffer; length & capacity are set to the string's
+ * length.
  */
 AZ_NODISCARD az_span az_span_from_str(char* str);
 
 /**
- * Creates an az_span literal which can be used to initialize a constant. For example
+ * @brief AZ_SPAN_LITERAL_FROM_BUFFER returns a literal az_span over an unitialized byte buffer. For
```

Comment:
What makes it an az_span literal are the rules of C. Here's an article I found: http://nickdesaulniers.github.io/blog/2013/07/25/designated-initialization-with-pointers-in-c/

Reply:
The buffer doesn't have to be uninitialized. If it is, use this macro and the length will be 0.
If it isn't use the other macro and the length will be the buffer's length. Sure, we can say this macro sets the length to 0.

Reply:
I would have thought `literal` meant something fixed or constant and not related to initializers and compact initialization syntax.

For instance, from https://www.tutorialspoint.com/cprogramming/c_constants.htm
> Constants refer to fixed values that the program may not alter during its execution. These fixed values are also called literals.

Even in the doc example, there is nothing preventing me from changing the characteristics of the span/contents of the buffer regardless of whether it was created as a literal or not. Is it really that helpful to encapsulate how the caller is going to use/call the span-creation method with specific macros? I am trying to understand the nuance a bit more here :)

For example, is it wrong for me to write this code:
```c
uint8_t buffer[] = {"Hello World"};
const az_span foo = AZ_SPAN_FROM_BUFFER(buffer); // instead of AZ_SPAN_LITERAL_FROM_BUFFER
```

And if we do want to keep the literal concept, should we call it `compound literal`, or is that implied already?
https://en.cppreference.com/w/c/language/compound_literal

> The buffer doesn't have to be uninitialized. If it is, use this macro and the length will be 0.
If it isn't use the other macro and the length will be the buffer's length.

Thank you for the clarification and link. That was quite helpful. Part of my intent with posing such questions was to suggest that maybe those types of questions should be answered/clarified in the API doc.

Reply:
A literal is NOT a constant. A literal is variable whose value to set/initialized at compile time. The value of the variable may change over time.

The code you wrote won't compile if foo is a global but it will compile if foo is a local. 

Literal is not our concept. It is a C language concept just like an expression is a C language concept. We have no choice but to keep these. We want to allow the customer code (and our code) to create global, initialized spans. And, we we also want to be able to create local, initialized spans. Unfortunately, the C compilers require different syntax to accomplish both these tasks and so, we have 2 macros.

The API doc and the conceptual docs I handed off to you have example of the usage for literal and expression macros. Between the docs and the sample, I think customers should get it. Plus, the customers know C, then they should be familiar with this too. And, if they get it wrong, the compiler will complain and then the customer should be able to figure this out without too much difficulty.